chips/lowrisc/src/uart.rs

```diff
@@ -0,0 +1,201 @@
+        txen OFFSET(0) NUMBITS(1) [],
+        txval OFFSET(1) NUMBITS(1) []
+    ],
+    val [
+        rx OFFSET(0) NUMBITS(16) []
+    ],
+    timeout_ctrl [
+        val OFFSET(0) NUMBITS(23) [],
+        en OFFSET(31) NUMBITS(1) []
+    ]
+];
+
+pub struct Uart<'a> {
+    registers: StaticRef<UartRegisters>,
+    clock_frequency: u32,
+    tx_client: OptionalCell<&'a dyn hil::uart::TransmitClient>,
+    rx_client: OptionalCell<&'a dyn hil::uart::ReceiveClient>,
+}
+
+#[derive(Copy, Clone)]
+pub struct UartParams {
+    pub baud_rate: u32,
+}
+
+impl Uart<'a> {
+    pub const fn new(base: StaticRef<UartRegisters>, clock_frequency: u32) -> Uart<'a> {
+        Uart {
+            registers: base,
+            clock_frequency: clock_frequency,
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+        }
+    }
+
+    fn set_baud_rate(&self, baud_rate: u32) {
+        let regs = self.registers;
+        let uart_ctrl_nco = ((baud_rate as u64) << 20) / self.clock_frequency as u64;
+
+        regs.ctrl
+            .write(ctrl::nco.val((uart_ctrl_nco & 0xffff) as u32));
+        regs.ctrl.modify(ctrl::tx.val(1 as u32));
+        regs.ctrl.modify(ctrl::rx.val(1 as u32));
+    }
+}
+
+impl hil::uart::UartData<'a> for Uart<'a> {}
+impl hil::uart::Uart<'a> for Uart<'a> {}
+
+impl hil::uart::Configure for Uart<'a> {
+    fn configure(&self, params: hil::uart::Parameters) -> ReturnCode {
+        let regs = self.registers;
+        // We can set the baud rate.
+        self.set_baud_rate(params.baud_rate);
+
+        regs.fifo_ctrl.write(fifo_ctrl::rxrst.val(1 as u32));
+        regs.fifo_ctrl.modify(fifo_ctrl::txrst.val(1 as u32));
+
+        // Disable interrupts as we don't suppor them
+        regs.intr_enable.set(0);
+
+        ReturnCode::SUCCESS
+    }
+}
+
+impl hil::uart::Transmit<'a> for Uart<'a> {
+    fn set_transmit_client(&self, client: &'a dyn hil::uart::TransmitClient) {
+        self.tx_client.set(client);
+    }
+
+    fn transmit_buffer(
+        &self,
+        tx_data: &'static mut [u8],
+        tx_len: usize,
+    ) -> (ReturnCode, Option<&'static mut [u8]>) {
+        let regs = self.registers;
+
+        if tx_len == 0 {
+            return (ReturnCode::ESIZE, Some(tx_data));
+        }
+
+        while regs.status.is_set(status::txfull) {}
+
+        // Fill the TX buffer until it reports full.
+        for i in 0..tx_len {
+            // Write the byte from the array to the tx register.
+            regs.wdata.write(wdata::data.val(tx_data[i] as u32));
+
+            // Check if the buffer is full and wait until it isn't
+            while regs.status.is_set(status::txfull) {}
+        }
+
+        // Wait for the send to complete
+        while regs.status.is_set(status::txempty) != true {}
+
+        (ReturnCode::SUCCESS, None)
+    }
+
+    fn transmit_abort(&self) -> ReturnCode {
+        ReturnCode::FAIL
+    }
+
+    fn transmit_word(&self, _word: u32) -> ReturnCode {
```

Comment:
Is a transmit_done callback issued? If not, this is problematic. You should implement transmit_word, or alternatively come up with a way to issue a callback. Note that you can't issue the callback from within the call, as this can create unsolvable race conditions.

Reply:
I'm not sure what you mean by a transmit_done callback.

I copied the lack of `transmit_word()` implementation from the HiFive1 UART :)

Reply:
Generally, Tock APIs are asynchronous and split-phase. A long-running I/O starts with a function call, which returns almost immediately (e.g., after setting up some registers to start the operation). When the operation completes, the implementation issues an asynchronous callback. It's called "split-phase" because these two phases -- before and after -- are split in execution. 

By transmit_done, I meant `transmitted_buffer`, which is implemented by a uart::TransmitClient. The normal operation of uart::Transmit is that a client calls Transmit::transmit_buffer, which starts or enqueues the transmission. Then, when it completes (e.g., in response to an interrupt), the implementation invokes uart::TransmitClient::transmitted_buffer to indicate the operation is over.

The nRF52 has an example of doing this with DMA: https://github.com/tock/tock/blob/master/chips/nrf52/src/uart.rs
The CC26x2 has an example of doing this byte-by-byte: https://github.com/tock/tock/blob/master/chips/cc26x2/src/uart.rs